ethereum-consensus/src/clock.rs

```diff
@@ -195,20 +195,25 @@ impl<T: TimeProvider + Send + Sync> Clock<T> {
 
 pub type SystemClock = Clock<SystemTimeProvider>;
 
+use tokio::time::interval_at;
 #[cfg(feature = "async")]
-use tokio_stream::Stream;
+use tokio_stream::{wrappers::IntervalStream, Stream, StreamExt};
 
 #[cfg(feature = "async")]
 impl<T: TimeProvider + Send + Sync> Clock<T> {
+    /// Returns a stream of slots. The stream will start from the next slot i.e it will not emit the
+    /// current slot.
     pub fn stream_slots(&self) -> impl Stream<Item = Slot> + '_ {
-        async_stream::stream! {
-            loop {
-                let slot = self.current_slot().expect("after genesis");
-                yield slot;
-                let duration_until_next_slot = self.duration_until_slot(slot + 1);
-                tokio::time::sleep(duration_until_next_slot).await;
-            }
-        }
+        let slot = self.current_slot().expect("after genesis");
+        let duration_until_next_slot = self.duration_until_slot(slot + 1);
+
+        // Stream will not emit the `current_slot`, it starts from the next slot. This is until a custom stream implementation is added to fix this. Ref: https://github.com/ralexstokes/ethereum-consensus/pull/281#discussion_r1350932968
+        let interval = interval_at(
+            tokio::time::Instant::now() + duration_until_next_slot,
+            Duration::from_secs(configs::mainnet::SECONDS_PER_SLOT),
```

Comment:
this should be parametric over the `Clock`s `SECONDS_PER_SLOT`

```suggestion
            Duration::from_secs(self.seconds_per_slot),
```